woocommerce/woocommerce
# CYS - Core:  Add async font installation

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR introduces an async job to install the fonts. 

The job starts when the `woocommerce_allow_tracking'` is set to yes and the Font Library is available (https://github.com/woocommerce/woocommerce/pull/44473/files#diff-9ff9cbf9fc5ce9eb0df48369788b45884e619f1841b94e4e3fce7bcb75c14e9fR104-R113).


The async job:
- gets the font info from the google font collection.
- create `wp_font_family` posts.
- download the various font_faces, upload them to the WordPress site, and create the `wp_font_face` posts.

This PR introduces two internal classes:
- `FontFace`
- `FontFamily` 

Both are utils classes to create and install the font families and font faces. As it is annotated in the code, we have to remove them and refactor this code when the WordPress PHP API to install fonts is available: https://github.com/WordPress/gutenberg/issues/58670


I think that we should add unit tests for the code that we added in this PR and in https://github.com/woocommerce/woocommerce/pull/44004 to ensure that we don't introduce any regression in the logic to install the fonts. What do you think?

Closes #44257 .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### These testing instructions require a connection with the database


1. Create a new WooCommerce installation with this version.
2. Ensure the WooCommerce Beta Tester plugin is installed and activated (available on this monorepo).
3. Head over to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` and enable `customize-store` feature flag.
4. Install this build of Gutenberg that contains the new Font Library API: [gutenberg.zip](https://github.com/woocommerce/woocommerce/files/14236920/gutenberg.zip)
5. Connect to the database and ensure that doesn't exist any post with these post-type: `wp_font_family`, `wp_font_face` and `wp_global_styles`.
6. Go on `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com` and enable the tracking.
7. Go on `/wp-admin/admin.php?page=wc-status&tab=action-scheduler` and ensure that `woocommerce_install_assembler_fonts` is in pending.
8. Ensure that the `woocommerce_install_assembler_fonts` job doesn't fail.
9. Visit the `wp-admin/admin.php?page=wc-admin&path=/customize-store`.
10. Open the network tools.
11. Follow the process.
12. Ensure that no POST requests are done on the `/font-faces/` endpoint. (it is expected that GET requests are done).
13. Wait until the Assembler is loaded.
14. Due [to this bug](https://github.com/WordPress/gutenberg/issues/58765), it is necessary to refresh the page to ensure that the fonts are loaded.
15. Click on "Change Fonts".
16. Ensure that the right fonts are applied.




### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
